torchvision/models/detection/ssd.py

```diff
@@ -0,0 +1,486 @@
+            i += 1
+        return out
+
+    def forward(self, x: List[Tensor]) -> Tensor:
+        all_results = []
+
+        for i, features in enumerate(x):
+            results = self._get_result_from_module_list(features, i)
+
+            # Permute output from (N, A * K, H, W) to (N, HWA, K).
+            N, _, H, W = results.shape
+            results = results.view(N, -1, self.num_columns, H, W)
+            results = results.permute(0, 3, 4, 1, 2)
+            results = results.reshape(N, -1, self.num_columns)  # Size=(N, HWA, K)
+
+            all_results.append(results)
+
+        return torch.cat(all_results, dim=1)
+
+
+class SSDClassificationHead(SSDScoringHead):
+    def __init__(self, in_channels: List[int], num_anchors: List[int], num_classes: int):
+        cls_logits = nn.ModuleList()
+        for channels, anchors in zip(in_channels, num_anchors):
+            cls_logits.append(nn.Conv2d(channels, num_classes * anchors, kernel_size=3, padding=1))
+        _xavier_init(cls_logits)
+        super().__init__(cls_logits, num_classes)
+
+
+class SSDRegressionHead(SSDScoringHead):
+    def __init__(self, in_channels: List[int], num_anchors: List[int]):
+        bbox_reg = nn.ModuleList()
+        for channels, anchors in zip(in_channels, num_anchors):
+            bbox_reg.append(nn.Conv2d(channels, 4 * anchors, kernel_size=3, padding=1))
+        _xavier_init(bbox_reg)
+        super().__init__(bbox_reg, 4)
+
+
+class SSD(nn.Module):
+    __annotations__ = {
+        'box_coder': det_utils.BoxCoder,
+        'proposal_matcher': det_utils.Matcher,
+    }
+
+    def __init__(self, backbone: nn.Module, anchor_generator: DBoxGenerator, size: Tuple[int, int], num_classes: int,
+                 image_mean: Optional[List[float]] = None, image_std: Optional[List[float]] = None,
+                 score_thresh: float = 0.01,
+                 nms_thresh: float = 0.45,
+                 detections_per_img: int = 200,
+                 iou_thresh: float = 0.5,
+                 topk_candidates: int = 400,
+                 positive_fraction: float = 0.25):
+        super().__init__()
+
+        # Use dummy data to retrieve the feature map sizes to avoid hard-coding their values
+        device = next(backbone.parameters()).device
+        tmp_img = torch.zeros((1, 3, size[1], size[0]), device=device)
+        tmp_sizes = [x.size() for x in backbone(tmp_img).values()]
+        out_channels = [x[1] for x in tmp_sizes]
+
+        assert len(out_channels) == len(anchor_generator.aspect_ratios)
+
+        self.backbone = backbone
+
+        self.anchor_generator = anchor_generator
+
+        self.box_coder = det_utils.BoxCoder(weights=(10., 10., 5., 5.))
+
+        # Estimate num of anchors based on aspect ratios: 2 default boxes + 2 * ratios of feaure map.
+        self.num_anchors = [2 + 2 * len(r) for r in anchor_generator.aspect_ratios]
+        self.head = SSDHead(out_channels, self.num_anchors, num_classes)
```

Comment:
I opted in for directly initialising the head here instead of passing it. Key issue with passing it is that we miss the `out_channels` info estimated above. Can be resolved by passing a callable, using a factory or doing the calculation of channels externally. This limitation needs to be addressed before implementing SSDlite.

Reply:
Agreed. As discussed offline, we will check the backbone to see if it has an attribute that provides the number of out_channels. If it does not, we will retrieve it via a utils method that can be replaced on the future using FX features. This way we will be able to both accept a head and initialize it if need be.